spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binder/PartitionHandler.java

```diff
@@ -0,0 +1,232 @@
+			DefaultBindingPropertiesAccessor properties, int partitionCount) {
+		this.beanFactory = beanFactory;
+		this.evaluationContext = evaluationContext;
+		this.partitionSelector = partitionSelector == null
+				? new DefaultPartitionSelector()
+				: partitionSelector;
+		this.partitioningMetadata = new PartitioningMetadata(properties, partitionCount);
+	}
+
+	/**
+	 * Return {@code true} if the binder properties provided indicate
+	 * that this binder is configured for partitioning.
+	 *
+	 * @return true if partitioning is enabled
+	 */
+	public boolean isPartitionedModule() {
+		return this.partitioningMetadata.isPartitionedModule();
+	}
+
+	/**
+	 * Determine the partition to which to send this message.
+	 * <p>
+	 * If a partition key extractor class is provided, it is invoked to determine
+	 * the key. Otherwise, the partition key expression is evaluated to obtain the
+	 * key value.
+	 * <p>
+	 * If a partition selector class is provided, it will be invoked to determine the
+	 * partition. Otherwise, if the partition expression is not null, it is evaluated
+	 * against the key and is expected to return an integer to which the modulo
+	 * function will be applied, using the {@code partitionCount} as the divisor. If no
+	 * partition expression is provided, the key will be passed to the binder
+	 * partition strategy along with the {@code partitionCount}. The default partition
+	 * strategy uses {@code key.hashCode()}, and the result will be the mod of that value.
+	 *
+	 * @param message the message.
+	 * @return the partition
+	 */
+	public int determinePartition(Message<?> message) {
+		Object key = null;
+		if (StringUtils.hasText(partitioningMetadata.partitionKeyExtractorClass)) {
+			key = invokeExtractor(partitioningMetadata.partitionKeyExtractorClass, message);
+		}
+		else if (partitioningMetadata.partitionKeyExpression != null) {
+			key = partitioningMetadata.partitionKeyExpression.getValue(this.evaluationContext, message);
+		}
+		Assert.notNull(key, "Partition key cannot be null");
+		int partition;
+		if (StringUtils.hasText(partitioningMetadata.partitionSelectorClass)) {
+			partition = invokePartitionSelector(partitioningMetadata.partitionSelectorClass,
+					key, partitioningMetadata.partitionCount);
+		}
+		else if (partitioningMetadata.partitionSelectorExpression != null) {
+			partition = partitioningMetadata.partitionSelectorExpression.getValue(this.evaluationContext, key, Integer.class);
+		}
+		else {
+			partition = this.partitionSelector.selectPartition(key, partitioningMetadata.partitionCount);
+		}
+		partition = partition % partitioningMetadata.partitionCount;
+		if (partition < 0) { // protection in case a user selector returns a negative.
+			partition = Math.abs(partition);
+		}
+		return partition;
+	}
+
+	private Object invokeExtractor(String partitionKeyExtractorClassName, Message<?> message) {
+		if (this.beanFactory.containsBean(partitionKeyExtractorClassName)) {
```

Comment:
I think that we do.

Furthermore, I'm not sure why we register the extractor with the `beanFactory` - could we just simply hold a reference to it and use that instead of going back to the `beanFactory` every time?